### DISCUSSION TITLE
Welcome to Blossom Discussions!

### DISCUSSION THREAD

**Discussion by ctopherngoc** [ctopherngoc/BlossomStory#113] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by JonesNMRN** [comment]
Hello Jones here, i am a big fan of Maplestory and always have been, i stumbled upon your project here on github whilst just browsing out of boredom, and i got super excited. This project is exactly what i have wanted to do but just haven't gotten started.

I am an indie game developer with about 12 years of experience in all sorts of engines.

One question: I have actually never used firebase, would you mind telling me what i would do in the project to get it setup for myself "locally" so that i can try to understand the project and maybe help with development or just to be able to develop it and share ideas?

Great work so far!

**Reply by ctopherngoc** [reply]
There's quite a few steps. It makes me think I should write me a readme file for it. I think you should be able to get it working with these steps:

1.  Download Godot 3.5 (this project does not work with Godot 4)
2. Create a Google Firebase project (record the project name)
3.  Navigate to Project Overview -> Project Settings  (record project API Key)
4. Setup Firebase Authentication and Firebase Database
5. In Firebase Authentication create two accounts GameServer and AuthenticationServer
6. In Firebase Database, create three collections: users, characters, items
![image](https://github.com/ctopherngoc/Blossom/assets/52055142/9819c286-c132-4f16-a325-692ec9fd2c45)

7. Create TWO files called server.json with the format:
![image](https://github.com/ctopherngoc/Blossom/assets/52055142/e5eb6b6b-94ac-4cdd-99dd-c24faeb4f639)

8. Enter information for GameServer in one of the .json file and save it in /server/data/server.json
9. Enter information for AuthServer for the other of the .json file and save it in /authentication/data/server.json
10. Create another server.json file with the following keys:
![image](https://github.com/ctopherngoc/Blossom/assets/52055142/afaf5960-1f7c-4729-96e6-c7eb4900949c)

11. Fill in json with your project info and local_ip. You can leave Server_IP blank or enter your local ip.
12. Save this server.json file in /client/data/server.json
13. You might have to generate a X509_Certificate and x509_key.key (you can run a similar gd.script as below)
![image](https://github.com/ctopherngoc/Blossom/assets/52055142/7e1bd09a-20ce-4ad3-b2db-660866597c35)

14. Place X509_Certificate in /client/resources/Certificate
15. Place X509_Certificate and x509_key.key in /gateway/Resources/Certificate
16.  Open four instances of Godot, one for GameServer, AuthServer, GatewayServer, and Client project
17. in client project -> global.gd set local: bool = true (shouldn't really matter if you have your local ip in both server and local in the server.josn)
18. Run the Servers in the following order Gateway -> Authentication -> GameServer -> Client

**Reply by JonesNMRN** [reply]
Thank you so much for the detailed explanation! I will try this out tonight or tomorrow, ill let you know how it goes.

**Comment by JonesNMRN** [comment]
I'm kinda stuck with the firebase web UI, is this the correct way to setup the database like you said or did i misunderstand something there?

![Skärmbild 2024-07-02 221748](https://github.com/ctopherngoc/Blossom/assets/87195463/102c1bd4-fce8-421d-b551-03ca02ff17f7)

**Reply by ctopherngoc** [reply]
You created a storage. You need to make a realtime database.

**Reply by JonesNMRN** [reply]
Yeah that's what i thought, it didn't feel right haha. Thanks!

**Reply by ctopherngoc** [reply]
I've updated the repository with the necessary files to run an instance of the game yourself. You can follow the directions in the readme.

all you really need to do is create a firebase account -> Setup auth and realtime database -> update the server.josn files

Hopefully it works for you!

**Reply by JonesNMRN** [reply]
Thank you so much for the detailed and fast explanation! I will test it out now.

**Reply by JonesNMRN** [reply]
Okay i got the hang of it now, the only thing is my realtime database does not look like yours. Do you have a detailed guide on how you set yours up in firebase?

**Reply by ctopherngoc** [reply]
Firebase possibly changed the name of the database. I took a look and there's two types: Cloud Firestore and Realtime Database. I have the Cloud Firebase version. Sorry for the misunderstanding. 

![image](https://github.com/ctopherngoc/Blossom/assets/52055142/c8bd7e72-ffee-4219-9dd9-dfa0d5394c62)

**Reply by ctopherngoc** [reply]
> Okay i got the hang of it now, the only thing is my realtime database does not look like yours. Do you have a detailed guide on how you set yours up in firebase?

Its pretty simple. In your firebase console, you want to go to the authentication section and create two accounts. One for your gameserver (so it can get data from the database) and one for the authentication server (so it can queue login information when login request occurs)

You take the server login and add it into the /server/data/server.json file. Similarly for the authentication login, you edit the server.json file in the /authentication/data/server.json

Files can be found in the most recent main of the repository in the local_host directory. Once the server.json files have been filled out correctly, you can drag and drop the client, server, authentication folders into the main repository folder and overwrite the files.

It should run and connect to your firebase authentication and firebase cloud

**Reply by JonesNMRN** [reply]
Haha don't worry! Thanks for the response again, now i stumble upon an error when starting the game server: 
![errorgodotserver](https://github.com/ctopherngoc/Blossom/assets/87195463/26fe9048-6584-49b5-9b74-d2e3f2715ec3)

And this is how my firestore database is setup: 
![cloudfirestore](https://github.com/ctopherngoc/Blossom/assets/87195463/7c338db4-07b6-4be7-ba37-66be217288ab)


Is there a specific way i have to setup the collections? like fields etc, if so how have you setup yours?

**Reply by ctopherngoc** [reply]
> Haha don't worry! Thanks for the response again, now i stumble upon an error when starting the game server: ![errorgodotserver](https://private-user-images.githubusercontent.com/87195463/345227165-26fe9048-6584-49b5-9b74-d2e3f2715ec3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk5NjY5MzYsIm5iZiI6MTcxOTk2NjYzNiwicGF0aCI6Ii84NzE5NTQ2My8zNDUyMjcxNjUtMjZmZTkwNDgtNjU4NC00OWI1LTliNzQtZDJlM2YyNzE1ZWMzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzAzVDAwMzAzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU3ZTIxNWY4NmMyMzJmZjEwNzg0ZTlhYjFkYWYzYmU0ODIxYzU0ZTk3MmIyMDAyZWNiYWUzODE0ZjRkN2E3NmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.H9P-GTJ5IGNN-lOlFs7F4SGO18VpEo11Vj6sdGq2DL8)
> 
> And this is how my firestore database is setup: ![cloudfirestore](https://private-user-images.githubusercontent.com/87195463/345227220-7c338db4-07b6-4be7-ba37-66be217288ab.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk5NjY5MzYsIm5iZiI6MTcxOTk2NjYzNiwicGF0aCI6Ii84NzE5NTQ2My8zNDUyMjcyMjAtN2MzMzhkYjQtMDdiNi00YmU3LWJhMzctNjZiZTIxNzI4OGFiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzAzVDAwMzAzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMzNTliN2E2NzZjMWE4YjgxY2M3YmI3MjVjYzRmOWFjZjIzNDAxYjE5OTRjODg5ZGFjNmY1NjljYjg3NTBhMjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.C39zTd4NZ6ywQaNZiQLje6GYRU56BPwA7iDcw_LxLcE)
> 
> Is there a specific way i have to setup the collections? like fields etc, if so how have you setup yours?

I just updated the repro to 4.0.2 that should fix your issue with the "invalid get index 'field'" All these issues you will be encountering is something I haven't experienced because I don't have an unofficial version of my game lol. The fix takes your project name from your server.json for some string manipulation instead of the hard coded one on line 256 of your screenshot.

I also fixed an issue with saving new items into the firebase dictionary for the server side.

**Reply by JonesNMRN** [reply]
I cloned the latest repo game server signs in successfully, but  then stops with "Invalid get index 'documents' (on base: 'Dictionary')."
0 - res://singletons/Firebase.gd:267 - at function: _server_get_document

**Reply by ctopherngoc** [reply]
![image](https://github.com/ctopherngoc/Blossom/assets/52055142/c903d675-8233-4e96-8210-81a970ee5f97)

I think this is the issue. Try adding this if statement

**Reply by JonesNMRN** [reply]
Thanks i got it working now. :D

Time to learn the code and see what i can come up with. :)

**Reply by ctopherngoc** [reply]
> Thanks i got it working now. :D
> 
> Time to learn the code and see what i can come up with. :)

Nice! Glad we can get it working for you. If you have any more trouble or questions please do ask. I'll try my best to answer it in a timely manner!

**Reply by JonesNMRN** [reply]
Is the combat just bricked at the moment? I am crashing sometimes when trying to attack, and i can't deal any damage.

**Reply by ctopherngoc** [reply]
> Is the combat just bricked at the moment? I am crashing sometimes when trying to attack, and i can't deal any damage.

The combat should work without issue (no crashing). It's server-authoritative combat so if you are not seeing damage numbers it might be your computer having issues running the seber and client simultaneously.

The server calculated the damage then communicates it to the client via world states with a timestamp. If there is lag the worldstete will be skipped to catch up.

What are the crashing messages you are getting?

**Reply by ctopherngoc** [reply]
I just took a look at it. It could possibly be your weapon type you have on the character. Go into your player data dictionary[equipment][rweapon][weaponType]. If its not a 1h_sword or 2h_sword, currently theres no attack animation for it on the server which is why you might not see damage numbers. I would edit it in firebase to be string 1h_sword and I'll resolve that in a bit

**Reply by ctopherngoc** [reply]
I've updated the repo to 4.1.0. Try to pull and rebase. Your combat issue should be fixed.

**Comment by JonesNMRN** [comment]
Hello sorry for the late response! Thank you!! I'll test it out soon, been busy with work but it will calm down after the weekend. :D
